llvm/include/llvm/DebugInfo/DWARF/DWARFContext.h

```diff
@@ -127,6 +127,9 @@ class DWARFContext : public DIContext {
 
   std::unique_ptr<const DWARFObject> DObj;
 
+  /// Binary containing debug information after objcopy --only-keep-debug
+  object::OwningBinary<object::ObjectFile> GnuLink;
```

Comment:
Could we avoid having this sort of invasive change by switching entirely to the debuglink binary early on & then the rest of the codepaths would work as they do today? (perhaps look at where the debuginfod is wired in - I would tihnk debuglink might work in a similar way?)

Reply:
I was aiming to do that here. This is the first entrance path for multiple tools that create DWARFContext. Otherwise this will need to be handled on caller side at multiple points. Unless I am missing something.
Where would be a better place for this?

Reply:
> Could we avoid having this sort of invasive change by switching entirely to the debuglink binary early on & then the rest of the codepaths would work as they do today? (perhaps look at where the debuginfod is wired in - I would tihnk debuglink might work in a similar way?)

OK poking around llvm source code looks like DebugInfoD is used on adhoc bases in various tools.
DebuginfodFetcher -- used in llvm-objdump, llvm-symbolizer, llvm-cov
Looking into APIs declared in Debuginfod.h used in LLDB, llvm-debuginfod-find

So doesn't look like there is a central place where it is "automatically" supported, and it's up to each tool to add support for it.

Reply:
@dwblaikie I realize in ::create is not super ideal. I was originally thinking about doing it in the constructor, but then if people go through ::create then DWARFObjInMemory is created for no reason.
Generally speaking I think this should be handled similar to .dwp where it's under the hood and consumers of debug information shouldn't really care/know if debug info is in main binary or in a separate elf object.

Reply:
Debuginfod (the "protocol") is actually quite underspecified for this scenario (I'm in the middle of adding full support for split, stripped, and "only-keep-debug" symbol acquisition through libdebuginfod) You *can't* replace the original binary with the .gnu.debuglink binary, because there's a high probability that the linked binary is missing a bunch of stuff from the stripped binary.

From what I've gathered they appear to be used as the "original split dwarf" because a common use is to make the .gnu.debuglink point to the an "objcopy --only-keep-debug" version of the binary, so it doesn't have anything *except* the DWARF data (there's almost no overlap between the --strip-debug and --only-keep-debug versions of the binary, and I *think* you need both do actually debug stuff...)

Reply:
As far as llvm's DWARF library is concerned, I /think/ we can replace the binary with the debuglink binary, though (even if it is only-keep-debug) - and that is what this patch is doing. Presumably that's what debuginfod support does too? (well, one way or another - as you say, perhaps underspecified for all these different flavors of split/partial binaries, etc - for now, I guess debuginfod support is retrieving the whole unstripped binary, so it's strictly better/more complete than the case being supported here with debuglink - but that points to, even more, that they could use the same path in this case (because the path is good enough for debuglink to only-keep-debug, so it's certainly fine to use that path for the more complete whole-unstripped-binary case for debuginfod))

Reply:
I feel like I'm missing something. As I understand it, replacing the stripped binary with an OKD binary would remove sections that are necessary for full symbolication, general debugging, and proper program execution (.symtab, .strtab, .shstrtab in my simple tests). If we replaced the stripped binary, how would we deal with those sections?

Reply:
> I feel like I'm missing something. As I understand it, replacing the stripped binary with an OKD binary would remove sections that are necessary for full symbolication, general debugging, and proper program execution (.symtab, .strtab, .shstrtab in my simple tests). If we replaced the stripped binary, how would we deal with those sections?

.symtab is actually copied
![image](https://github.com/llvm/llvm-project/assets/43973793/c81040e7-286b-48f1-a2a0-3f75c8b7c836)

That being said the .eh_frame is not. :/

Reply:
Looking more closely through DWARFContext.cpp except for dump in DWARFContext getting DwarfObject is controlled through getDWARFObj(). 
One alternative is in DWARFContext to have both gnulink one, and main binary one. Depending if gnulink is present in it return either it or main binary DWARFObject. With a default parameter in getDWARFObj that can override behavior for apis that need it. Like when eh_frame is used.

This way code can also be moved into constructor.

Reply:
> Presumably that's what debuginfod support does too? (well, one way or another - as you say, perhaps underspecified for all these different flavors of split/partial binaries, etc - for now, I guess debuginfod support is retrieving the whole unstripped binary, so it's strictly better/more complete than the case being supported here with debuglink - but that points to, even more, that they could use the same path in this case (because the path is good enough for debuglink to only-keep-debug, so it's certainly fine to use that path for the more complete whole-unstripped-binary case for debuginfod))

What I ended up doing for debuginfod is kind-of situational; generally the tools that can invoke debuginfod look for debug info in the binaries that are directly available to them, but failing that, they retrieve the debug binary and then use it just for its DWARF. For tools that only operate on DWARF, the fetched binary can replace the original, but for other tools, the DWARF that the tool analyzes may be loaded from a different binary than the one the rest of the tool operates on.

Last time I checked (and if I understand their code correctly), the GNU implementation handles this in a more centralized way; both debuglink sections and debuginfod lookups can provide additional auxiliary sections that get tacked on to a binary. (Unfortunately I don't remember the precise semantics of this.) Something that can merge sections like this would be the best of both worlds, since it would produce an uber-image that could be passed around trivially. But I didn't see a straightforward way to do this in LLVM.